File: frontend/src/mentions.js
export function mentionRecipient(text, models) {
  const value = text.trimStart();
  if (!value.startsWith("@")) return null;
  const matches = models.flatMap((model) =>
    [model.name, model.id].flatMap((alias) =>
      [alias, `"${alias}"`]
        .filter((label) => {
          const prefix = `@${label}`;
          return (
            value.toLowerCase().startsWith(prefix.toLowerCase()) &&
            (value.length === prefix.length ||
              /[\s:,]/.test(value[prefix.length]))
          );
        })
        .map((label) => ({ model, length: label.length + 1 })),
    ),
  );
  if (!matches.length)
    return { error: "Choose an @model from this project's suggestions." };
  const longest = Math.max(...matches.map((m) => m.length));
  const found = [
    ...new Map(
      matches.filter((m) => m.length === longest).map((m) => [m.model.id, m]),
    ).values(),
  ];
  if (found.length !== 1)
    return {
      error: "This model name is ambiguous. Choose its unique @profile ID.",
    };
  const { model, length } = found[0];
  if (!model.enabled)
    return { error: "This model is disabled. Enable it in Project setup." };
  return { model, empty: !value.slice(length).replace(/^[\s:,]+/, "") };
}

File: backend/mother/mentions.py
"""Resolve an explicit leading @recipient before dispatching any model calls."""

import re


def recipient(content, models):
    text = content.lstrip()
    if not text.startswith("@"):
        return None
    matches = []
    for model in models:
        for alias in (model["name"], model["id"]):
            for label in (alias, f'"{alias}"'):
                prefix = "@" + label
                if text.casefold().startswith(prefix.casefold()) and (
                    len(text) == len(prefix) or re.match(r"[\s:,]", text[len(prefix)])
                ):
                    matches.append((len(prefix), model))
    if not matches:
        raise ValueError("Unknown @model. Choose a model from this project's suggestions.")
    length = max(size for size, _ in matches)
    found = {model["id"]: model for size, model in matches if size == length}
    if len(found) != 1:
        raise ValueError("Ambiguous @model name. Use the unique @profile ID instead.")
    model = next(iter(found.values()))
    if not model["enabled"]:
        raise ValueError("This @model is disabled. Enable it in Project setup first.")
    if not text[length:].lstrip(" \t\r\n:,"):
        raise ValueError("Write a message after the @model name.")
    return model["id"]
